homeassistant/components/device_tracker/icloud.py

```diff
@@ -47,39 +50,39 @@ def keep_alive(now):
         api.authenticate()
         _LOGGER.info("Authenticate against iCloud")
 
-    track_utc_time_change(hass, keep_alive, second=0)
-
     def update_icloud(now):
         """Authenticate against iCloud and scan for devices."""
         try:
             # The session timeouts if we are not using it so we
             # have to re-authenticate. This will send an email.
-            api.authenticate()
+            keep_alive(None)
             # Loop through every device registered with the iCloud account
             for device in api.devices:
                 status = device.status()
                 location = device.location()
                 # If the device has a location add it. If not do nothing
                 if location:
                     see(
-                        dev_id=re.sub(r"(\s|\W|')",
-                                      '',
-                                      status['name']),
+                        dev_id=slugify(status['name'].replace(' ', '', 99)),
                         host_name=status['name'],
                         gps=(location['latitude'], location['longitude']),
                         battery=status['batteryLevel']*100,
                         gps_accuracy=location['horizontalAccuracy']
                     )
-                else:
-                    # No location found for the device so continue
-                    continue
         except PyiCloudNoDevicesException:
             _LOGGER.info('No iCloud Devices found!')
 
-    track_utc_time_change(
-        hass, update_icloud,
-        minute=range(0, 60, config.get(CONF_INTERVAL, DEFAULT_INTERVAL)),
-        second=0
-    )
+    hass.bus.listen_once(EVENT_HOMEASSISTANT_START, update_icloud)
+
+    interval = config[CONF_INTERVAL]
+    track_utc_time_change(hass, update_icloud, second=0,
+                          minute=range(0, 60, interval))
+
+    # Schedule keepalives between the updates
+    keep_alive_start = KEEPALIVE_INTERVAL
```

Comment:
Maybe you should take a look at my custom version of the icloud tracker, I changed it to have a dynamic interval for each device instead of having a fixed interval for all device for one account

Op 25 jul. 2016 om 07:28 heeft Johann Kellerman <notifications@github.com<mailto:notifications@github.com>> het volgende geschreven:


In homeassistant/components/device_tracker/icloud.py<https://github.com/home-assistant/home-assistant/pull/2614#discussion_r72010135>:

>          except PyiCloudNoDevicesException:
>              _LOGGER.info<http://LOGGER.info>('No iCloud Devices found!')
>
> -    track_utc_time_change(
> -        hass, update_icloud,
> -        minute=range(0, 60, config.get(CONF_INTERVAL, DEFAULT_INTERVAL)),
> -        second=0
> -    )
> +    hass.bus.listen_once(EVENT_HOMEASSISTANT_START, update_icloud)
> +
> +    interval = config[CONF_INTERVAL]
> +    track_utc_time_change(hass, update_icloud, second=0,
> +                          minute=range(0, 60, interval))
> +
> +    # Schedule keepalives between the updates
> +    keep_alive_start = KEEPALIVE_INTERVAL


Because of keep_alive_start += KEEPALIVE_INTERVAL (KEEPALIVE_INTERVAL=4)

If interval = 8 it should only be
4, 12, 20, 28, 36, 44, 52

If interval=9 there will be a two
4, 12, 20, 28, 36, 44, 52
8, 16, 24, 32, 40, 48, 56

Chances are small we'd have a second (except if interval becomes very big), but will rather combine and ensure only one gets registered

-
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/home-assistant/home-assistant/pull/2614/files/99e5954b9b4530b69c838cfc0617c9895162c7ec#r72010135>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABERpmDyxmeztjdcDzBBYTagZI5gbuGFks5qZEl6gaJpZM4JTmyj>.